[RFC] vim-patch:7.4.1547

scampi opened the pull request:
#### vim-patch:7.4.1547

Problem:    Getting a cterm highlight attribute that is not set results in the
            string "-1".
Solution:   Return an empty string. (Taro Muraoka)

https://github.com/vim/vim/commit/385111bd86e0b38667879c3e89506ca1ae98e1df

scampi committed:
vim-patch:7.4.1547

Problem:    Getting a cterm highlight attribute that is not set results in the
            string "-1".
Solution:   Return an empty string. (Taro Muraoka)

https://github.com/vim/vim/commit/385111bd86e0b38667879c3e89506ca1ae98e1df

justinmk commented:
Thanks! `test/functional/terminal/highlight_spec.lua` needs to be updated (separate commit, same PR) since it tests for `-1` result.

justinmk commented on src/nvim/syntax.c:
```diff
@@ -7003,6 +7003,8 @@ highlight_color (
     else
       n = HL_TABLE()[id - 1].sg_cterm_bg - 1;
+    if (n < 0)
+      return NULL;
```
in a separate "lint" commit, please add `{...}` around this block. The linter "cascades", so it will also check up to line 7002 (just add `{...}` to those blocks, too)

scampi committed:
synIDattr now returns an empty string when the cterm attribute is not set

scampi committed:
lint

src/nvim/syntax.c:7006:  if should always use braces  [readability/braces] [5]
src/nvim/syntax.c:7008:  Never use sprintf.  Use snprintf instead.  [runtime/printf] [5]

scampi commented:
@justinmk running the linter caught another error about `sprintf`

scampi commented:
Running `make test` on my system blocks at some point:

```sh
$ LC_ALL=C make test
ninja  -C .deps
ninja: Entering directory `.deps'
[2/2] cd /home/yfful/documents/code/neovim/.deps && /usr/bin/cmake -E touch .third-party
ninja  -C build
ninja: Entering directory `build'
[2/2] cd /home/yfful/documents/code/neovim/build/runtime && /usr/bin/cmake -E copy_directory /h...ocuments/code/neovim/build/bin/nvim -u NONE -i NONE -e --headless -c helptags\ ++t\ doc -c quit
ninja  -C build functionaltest
ninja: Entering directory `build'
[0/1] cd /home/yfful/documents/code/neovim/build && /usr/bin/cmake -DBUSTED_PRG=/home/yfful/doc...T_TYPE=functional -DSYSTEM_NAME=Linux -P /home/yfful/documents/code/neovim/cmake/RunTests.cmake
◌●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●◌●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●◌●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●◌●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●◌●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●
1678 successes / 0 failures / 0 errors / 5 pending : 66.962623 seconds

Pending → ...documents/code/neovim/test/functional/viml/lang_spec.lua @ 11
viml parses `<SID>` with turkish locale
...documents/code/neovim/test/functional/viml/lang_spec.lua:11: Locale tr_TR.UTF-8 not supported

Pending → ...nts/code/neovim/test/functional/ui/screen_basic_spec.lua @ 566
Screen resize has minimum width/height values

Pending → ...nts/code/neovim/test/functional/terminal/cursor_spec.lua @ 63
terminal cursor with number column is positioned correctly when focused

Pending → ...ments/code/neovim/test/functional/provider/ruby_spec.lua @ 20
Cannot find the neovim RubyGem. Try :CheckHealth

Pending → .../documents/code/neovim/test/functional/core/job_spec.lua @ 425
jobs will only emit the "exit" event after "stdout" and "stderr"
```

justinmk commented:
`make test` does not run the legacy Vim tests (like e.g. the one you added in this PR, which appears to be [failing in the build](https://s3.amazonaws.com/archive.travis-ci.org/jobs/159118419/log.txt)); try `make oldtest` instead.

Just for reference, if you want to see what's causing `make test` to fail, change `utfTerminal` to `gtest` here: https://github.com/neovim/neovim/blob/master/CMakeLists.txt#L409

scampi commented:
The test I added checks the `bold` attribute of the `term` mode https://github.com/neovim/neovim/pull/5326/files#diff-b718614c01ecee629344867a1a86daf0R7

However, the [highlight_has_attr](https://github.com/neovim/neovim/blob/master/src/nvim/syntax.c#L6926) method does not have the `term` mode. In fact, `sg_term` is missing completely although it does appear in [vim](https://github.com/vim/vim/blob/master/src/syntax.c#L8997). Any reason why the `sg_term` mode is missing from the `hl_group` struct ?

scampi commented:
Also, do you have any tips on how to debug tests like the one I added ? From the wiki, the mentions of GDB are only about functional tests or a running nvim instance, but nothing about an oldtest.

justinmk commented:
Isn't `term` deprecated? We can just remove those checks, maybe

scampi commented:
@justinmk there are mentions in the doc about `term`, e.g., https://github.com/neovim/neovim/blob/7eb4d2f79dcc712dae1513516b9db5f574d51437/runtime/doc/eval.txt#L6754
I can try fixing the doc in another PR ?

scampi committed:
removed deprecated tests of the term highlight mode

scampi commented:
Something wrong with the folding, the indent is expected to be 2 but was 0:

```
test/functional/legacy/045_folding_spec.lua:97: Expected objects to be the same.

Passed in:
(string) 'manual 1 aa
-1
3 cc
1 aa
marker 2
1
0
0
indent 0
0
folding 9 ii
    3 cc
7 gg
8 hh
expr 2
1
2
0'
Expected:
(string) 'manual 1 aa
-1
3 cc
1 aa
marker 2
1
1
0
indent 2
1
folding 9 ii
    3 cc
7 gg
8 hh
expr 2
1
2
0'
stack traceback:
	...eovim/neovim/test/functional/legacy/045_folding_spec.lua:97: in function
	<...eovim/neovim/test/functional/legacy/045_folding_spec.lua:10>
```

justinmk commented:
Please ignore the fold failure, that's an issue on master.

Shougo commented:
@scampi It is my fault.  I will fix it.

justinmk commented:
Merged. Thanks!

scampi commented:
cool! thanks @Shougo
